designs/operator-monitoring-plugin.md

```diff
@@ -0,0 +1,209 @@
+---
+title: controller-observation-plugin
+authors:
+  - "@Kavinjsir"
+reviewers:
+  - "@varsha"
+  - "@camila"
+approvers:
+creation-date: 2022-06-06
+last-updated: 2022-06-06
+status: provisional
+see-also:
+  - ""
+replaces:
+  - ""
+superseded-by:
+  - ""
+---
+
+# a-grafana-dashboard-plugin-to-observe-controller-metrics
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Open Questions [optional]
+
+- TBD
+
+## Summary
+
+This EP will provide a way for the operator author to visualize the runtime metrics in order to observe the operator status.
+
+The functionality will be implemented by a new plugin that can generate the 'manifests' to be loaded on Grafana.
+
+In the context of this EP, the Grafana 'manifests' are mainly:
+
+- JSON files to be directly loaded on Grafana dashboards
+- TBD
+
+## Motivation
+
+The [controller-runtime](https://github.com/kubernetes-sigs/controller-runtime) provides rich metrics to measure controllers' status from various perspectives.
+
+Currently, users of Kubebuilder need to design and create their own Grafana dashboards to visualize the metrics.
+It would be great if we can provide a unified and comprehensive approach so that users may:
+
+- easily visualize controller/operator metrics
+- avoid duplicate creation of Grafana dashboards file
+- improve the observability towards the opeartor, keep it safe and robust
+
+### Goals
+
+- Provide manifests that can display operator status for common needs.
+- Add a plugin or enhance the current kustomize plugin that scaffolds Grafana manifests.
+
+### Non-Goals
+
+- TBD
+
+## Proposal
+
+This proposed implementation is to create a new plugin, `grafana`, that scaffolds the manifests which are loaded by Grafana to display dashboards for controllers' status.
+
+Basically, the manifest can be raw json file such as `controller-dashboard.json` that can be directly applied to Grafana Web UI.
```

Comment:
Is required to be a JSON file OR could we have the manifest with the dashboard?
So that, users would only apply the manifests by command line and check the dashboards with graphs and end etc on the UI directly? 

c/c @Kavinjsir @varshaprasad96

Reply:
The JSON file provides the [dashboard](https://grafana.com/docs/grafana/latest/dashboards/json-model/) template, (ie) UI related. How do we apply it on cluster? Could we elaborate on that? In the sense, do the users have to just import this JSON specified dashboard in their grafana instance or is there any other way to do so?

Reply:
For this json file, the basic approach is to let the user **copy** the content and **paste** it in their Grafana Web UI.

So this would assume:
1. The users have the write access to their grafana service
2. The users' Grafana service allows dashboards creation and storage from Web UI.

Reply:
The above idea may not be ideal. It assumes users are commonly doing dashboards creation work on Web UI directly.

I know for `prometheus-operator` (or `kube-prometheus`), the json content can be written in a `K8s ConfigMap`. That means:
- The dashboard is installed when the configmap is applied to the cluster.
- The installation process can be handled in Makefile with a simple command.

I personally like this approach, but I'm not sure how common is `prometheus-operator` or `kube-prometheus`.

Or can we assume people would use these stacks?

Reply:
In other case, like some people who are familiar with the Grafana community, they may setup their monitoring stacks through jsonnet. The grafana community provides [tanka](https://tanka.dev/) to provision k8s resources that can operate prometheus and grafana.

In that case, it maybe hard to install the dashboard at our side. So this maybe another story.

Reply:
So it seems challenging to provide a general approach to fit all cases. The dashboard installation process seems depends on how users deploy their Grafana.

This is also the reason why I come back to the very basic way as just scaffolding out the raw json file.

Reply:
Would you have some ideas or suggestions on this?